Source/AppAuthTV/OIDTVTokenRequest.m

```diff
@@ -0,0 +1,153 @@
+                         codeVerifier:(nullable NSString *)codeVerifier
+                 additionalParameters:
+                     (nullable NSDictionary<NSString *, NSString *> *)additionalParameters
+    OID_UNAVAILABLE_USE_INITIALIZER(@selector
+                                     (initWithConfiguration:
+                                                   grantType:
+                                                  deviceCode:
+                                                    clientID:
+                                                clientSecret:
+                                        additionalParameters:
+                                              ))
+
+- (instancetype)initWithConfiguration:(OIDTVServiceConfiguration *)configuration
+                            grantType:(NSString *)grantType
+                           deviceCode:(NSString *)deviceCode
+                             clientID:(NSString *)clientID
+                         clientSecret:(NSString *)clientSecret
+                 additionalParameters:(NSDictionary<NSString *, NSString *> *)additionalParameters {
+  self = [super initWithConfiguration:configuration
+                            grantType:grantType
+                    authorizationCode:nil
+                          redirectURL:[[NSURL alloc] initWithString:@""]
+                             clientID:clientID
+                         clientSecret:clientSecret
+                                scope:nil
+                         refreshToken:nil
+                         codeVerifier:nil
+                 additionalParameters:additionalParameters];
+
+  if (self) {
+    _deviceCode = [deviceCode copy];
+  }
+  return self;
+}
+
+#pragma mark - NSCopying
+
+- (instancetype)copyWithZone:(nullable NSZone *)zone {
+  // The documentation for NSCopying specifically advises us to return a reference to the original
+  // instance in the case where instances are immutable (as ours is):
+  // "Implement NSCopying by retaining the original instead of creating a new copy when the class
+  // and its contents are immutable."
+  return self;
+}
+
+#pragma mark - NSSecureCoding
+
++ (BOOL)supportsSecureCoding {
+  return YES;
+}
+
+- (instancetype)initWithCoder:(NSCoder *)aDecoder {
+  self = [super initWithCoder:aDecoder];
+  if (self) {
+    NSString *deviceCode = [aDecoder decodeObjectOfClass:[NSString class] forKey:kDeviceCodeKey];
+    _deviceCode = deviceCode;
+  }
+  return self;
+}
+
+- (void)encodeWithCoder:(NSCoder *)aCoder {
+  [super encodeWithCoder:aCoder];
+  [aCoder encodeObject:_deviceCode forKey:kDeviceCodeKey];
+}
+
+- (OIDURLQueryComponent *)tokenRequestBody {
+  OIDURLQueryComponent *query = [[OIDURLQueryComponent alloc] init];
+
+  if (self.grantType) {
+    [query addParameter:kGrantTypeKey value:self.grantType];
+  }
+
+  if (self.deviceCode) {
+    [query addParameter:kDeviceCodeKey value:self.deviceCode];
+  }
+
+  [query addParameters:self.additionalParameters];
+
+  return query;
```

Comment:
Shouldn't you add the rest of the parameters (at least clientID)?

Reply:
Client ID is added in the URLRequest code in the parent; it will add clientID then if not doing http basic authentication ( https://github.com/openid/AppAuth-iOS/blob/master/Source/AppAuthCore/OIDTokenRequest.m#L277 )